py/sentry_streams/window.py

```diff
@@ -0,0 +1,108 @@
+from dataclasses import dataclass
+from datetime import timedelta
+
+
+@dataclass
+class Trigger:
+    """
+    A generic representation of a window trigger.
+    The trigger defines when a window is marked
+    as completed in order to fire a function
+    or computation on the window.
+    """
+
+
+@dataclass
+class EventTimeTrigger(Trigger):
+    """
+    Triggers based on progress of
+    event time of event data. Once the
+    watermark exceeds the end of the
+    window, window function is triggered.
+    """
+
+
+@dataclass
+class CountingTrigger(Trigger):
+    """
+    Specifically triggers window functions
+    (for example, aggregates) when the number
+    of elements in a window exceeds the count.
+    """
+
+    count: int
```

Comment:
Isn't the type of trigger coupled to the type you have in the Window (datetime vs count)?
In other words, I'd assume we cannot use an IntervalTrigger in a SlidingCountWindow.

If I am correct, I think we should force a specific type of window to have a specific type of trigger.
Can we do this with type hints maybe making the Trigger generic with respect to the type used inside and force a window adn trigger to use the same type when the type hint is resolved?

Reply:
I think there are cases in which the two don't have to match, for example:
- Global windows. These will window all data seen if they have the same key. You can apply a variety of Triggers to this. 
- Your condition for closing windows is different from the condition for triggering a computation. What if you want to window data by their event times (and you could have late data), but you want to trigger computation as soon as you have X events in the window? (Something like an alert threshold being exceeded?)

Flink also allows you to plug in/customize triggers with whatever windowing mechanism you have. 

Don't know if we want to anticipate this level of flexibility already. WDYT?

Reply:
Discussed offline:

since we don't have any use cases in mind which will need a trigger that's different from the window type, we can just remove triggers altogether for now.